src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,541 @@
+	{
+		strlcpy(key.database, get_database_name(MyDatabaseId), NAMEDATALEN);
+	}
+
+	if (CurrentCoordinatedTransactionState == COORD_TRANS_NONE)
+	{
+		CurrentCoordinatedTransactionState = COORD_TRANS_IDLE;
+	}
+
+	/*
+	 * Lookup relevant hash entry. We always enter. If only a cached
+	 * connection is desired, and there's none, we'll simply leave the
+	 * connection list empty.
+	 */
+
+	entry = hash_search(ConnectionHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		entry->connections = NIL;
+	}
+
+	if (flags & CACHED_CONNECTION)
+	{
+		ListCell *connectionCell = NULL;
+
+		/* check connection cache for a connection that's not already in use */
+		foreach(connectionCell, entry->connections)
+		{
+			connection = (MultiConnection *) lfirst(connectionCell);
+
+			/* don't return claimed connections */
+			if (!connection->claimedExclusively)
+			{
+				if (flags & SESSION_LIFESPAN)
+				{
+					connection->sessionLifespan = true;
+				}
+				connection->activeInTransaction = true;
+
+				/*
+				 * Check whether we're right now allowed to open new
+				 * connections. A cached connection counts as new if it hasn't
+				 * been used in this transaction.
+				 *
+				 * FIXME: This should be removed soon, once all connections go
+				 * through this API.
+				 */
+				if (!connection->activeInTransaction &&
+					XactModificationLevel > XACT_MODIFICATION_DATA)
+				{
+					ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+									errmsg("cannot open new connections after the first "
+										   "modification command within a transaction")));
+				}
+
+				return connection;
+			}
+
+			/*
+			 * One could argue for erroring out when the connection is in a
+			 * failed state. But that'd be a bad idea for two reasons:
+			 *
+			 * 1) Generally starting a connection might fail, after calling
+			 *    this function, so calling code needs to handle that anyway.
+			 * 2) This might be used in code that transparently handles
+			 *    connection failure.
+			 */
+		}
+
+		/* no connection available, done if a new connection isn't desirable */
+		if (!(flags & NEW_CONNECTION))
+		{
+			return NULL;
+		}
+	}
+
+	/*
+	 * Check whether we're right now allowed to open new connections.
+	 *
+	 * FIXME: This should be removed soon, once all connections go through
+	 * this API.
+	 */
+	if (XactModificationLevel > XACT_MODIFICATION_DATA)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("cannot open new connections after the first modification "
+							   "command within a transaction")));
+	}
+
+	/*
+	 * Either no caching desired, or no pre-established, non-claimed,
+	 * connection present. Initiate connection establishment.
+	 */
+	connection = StartConnectionEstablishment(&key);
+
+	oldContext = MemoryContextSwitchTo(ConnectionContext);
+	entry->connections = lappend(entry->connections, connection);
+	MemoryContextSwitchTo(oldContext);
+
+	if (flags & SESSION_LIFESPAN)
+	{
+		connection->sessionLifespan = true;
```

Comment:
For simplicity, requiring to caller to think about whether a connection should have a session lifespan seems undesirable. It currently happens to work because the router executor doesn't claim connections exclusively, uses appropriate caching flags, and therefore under normal conditions either always reuses a connection or opens a new one using session lifespan, and there are no other callers that use the flag, meaning in the end we cache 1 session lifespan connection. That's a lot of stuff to have to understand to know how we cache connections and likely to be broken at some point if we force that complexity onto callers.

It'd be nicer if we hide that complexity by implementing a simple caching policy, e.g.: "the first connection to a node always gets sessionLifespan" and revising it when the need arises.